File: Javascript/day06/day-06-b.js
const fs = require("fs");

const input = fs.readFileSync("./input.txt", "utf8");

FirstStartOfMessageMarker(input, 14);

function FirstStartOfMessageMarker(dataStream, numOfDistinctChars) {
  let chunks = windows(dataStream, numOfDistinctChars);

  for (let i = 0; i < chunks.length; i++) {
    const chunk = chunks[i];
    if (isAllUnique(chunk, numOfDistinctChars)) {
      console.log(i + numOfDistinctChars);
      break;
    }
  }
}

/**
 *
 * @param {string} arr
 * @param {number} windowSize
 * @returns {string[]}
 */
function windows(input, windowSize) {
  let result = [];

  for (let i = 0; i <= input.length - windowSize; i++) {
    result.push(input.slice(i, i + windowSize));
  }

  return result;
}

/**
 *
 * @param {string} chunk
 * @returns {boolean}
 */
function isAllUnique(chunk, numOfDistinctChars) {
  let uniqueChars = [];

  for (let ch of chunk) {
    if (!uniqueChars.includes(ch)) {
      uniqueChars.push(ch);
    }
  }

  return uniqueChars.length == numOfDistinctChars;
}

File: Go/day06/day-06-b.go
package main

import (
	"fmt"
	"os"
	"strings"
)

func main() {
	bytes, _ := os.ReadFile("./input.txt")
	input := string(bytes)

	FirstStartOfMessageMarker(input, 14)
}


func FirstStartOfMessageMarker(dataStream string, numOfDistinctChars int) {
	chunks := windows(dataStream, numOfDistinctChars)

	for i, chunk := range chunks {
		if isAllUnique(chunk, numOfDistinctChars) {
			fmt.Println(i + numOfDistinctChars)
			break
		}
	}
}

func windows(input string, windowSize int ) []string {
	chunks := []string{}

	for i := 0 ; i<= len(input) - windowSize; i++ {
		chunks = append(chunks, input[i: i+windowSize])
	}

	return chunks
}

func isAllUnique(input string, numOfDistinctChars int) bool {
	uniqueChars := ""

	for _, ch := range input {
		if !strings.ContainsRune(uniqueChars, ch) {
			uniqueChars += string(ch)
		}
	}

	return len(uniqueChars) == numOfDistinctChars
}